### PR TITLE
[processor/cumulativetodelta] make first point drop configurable

### LINKED ISSUE #20770: [processor/cumulativetodelta] make first point drop configurable
### Component(s)

processor/cumulativetodelta

### Is your feature request related to a problem? Please describe.

PR #18224 unconditionally drops the first first point in converting a metric from cumulative to delta temporality, using the initial point to establish a baseline for future deltas.
Somewhat unsurprisingly, we have received feedback that it drops too much data, 
especially if an instance has started due to a scale up (so the initial point contains a significant value).

Thinking about it some more, the need to drop the first point comes from unaligned service instance lifecycles: a "first" point is observed on the start of the application or the collector.

I think the impact of trying to do an out of application cumulative to delta conversion may be reduced by running the collector as a sidecar, tying its lifecycle to that of the application, removing the need to drop first points.

### Describe the solution you'd like

A config field similar to [awsemfexporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/awsemfexporter)'s
`retain_initial_value_of_delta_metric` setting that controls where the initial data point is dropped.

### Describe alternatives you've considered

give the processor a file to persist state across restarts

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by seankhliao**
**Description:**

Make the first handling of the first observed point configurable to be applicable in different usecases.

**Link to tracking Issue:** ##20770

**Testing:** updated tests

**Documentation:** Readme added

### CODE REVIEW

File: processor/cumulativetodeltaprocessor/internal/tracking/tracker.go
Comment (seankhliao) [inline processor/cumulativetodeltaprocessor/internal/tracking/tracker.go:144]: I'm a bit unsure about the correct value for start time if we unconditionally keep the value

File: processor/cumulativetodeltaprocessor/README.md
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/README.md:27]:
```suggestion
  - `auto`: send if and only if the startime is set AND the starttime happens after the component started AND the starttime is different from the timestamp.
```

as much as I love the `iff` notation, probably best to write the whole thing out in case people are unfamiliar with it.

File: processor/cumulativetodeltaprocessor/README.md
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/README.md:26]: Can you add to the README why someone might choose `drop` vs `keep` vs `auto`.

File: processor/cumulativetodeltaprocessor/config.go
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/config.go:34]: I believe we need to update https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6eb5eff7ce198c56b82903cab7ee43729ba5af64/processor/cumulativetodeltaprocessor/factory.go#L39 in order for `auto` to be used as default.

File: processor/cumulativetodeltaprocessor/config_test.go
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/config_test.go:1]: The existing tests that don't error should be updated to check that `initial_value` is set to `auto`.

File: processor/cumulativetodeltaprocessor/config.go
Comment (seankhliao) [reply processor/cumulativetodeltaprocessor/config.go:34]: I don't think so, auto was defined as the first value `= iota`, with the real value of `0`, so it will be default if we don't set anything

File: processor/cumulativetodeltaprocessor/README.md
Comment (seankhliao) [reply processor/cumulativetodeltaprocessor/README.md]: I've added some guidance.

File: processor/cumulativetodeltaprocessor/config_test.go
Comment (seankhliao) [reply processor/cumulativetodeltaprocessor/config_test.go:1]: done

File: processor/cumulativetodeltaprocessor/internal/tracking/tracker.go
Comment (seankhliao) [reply processor/cumulativetodeltaprocessor/internal/tracking/tracker.go]: ah right, fixed
also the one below which i can't comment on

File: processor/cumulativetodeltaprocessor/internal/tracking/tracker.go
Comment (seankhliao) [reply processor/cumulativetodeltaprocessor/internal/tracking/tracker.go]: alright, i've removed the timestamp

**Review by mx-psi — APPROVED** [review]

**Comment by mx-psi** [comment]
@TylerHelmuth will let you merge this one once you have given it a final review :)

File: processor/cumulativetodeltaprocessor/internal/tracking/tracker.go
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/internal/tracking/tracker.go:156]: Can we move this inside the `!ok` if condition after the switch statement?

File: processor/cumulativetodeltaprocessor/internal/tracking/tracker.go
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/internal/tracking/tracker.go:162]: I don't believe it is possible to reach this condition unless `ok` is `true`.

File: processor/cumulativetodeltaprocessor/internal/tracking/tracker.go
Comment (seankhliao) [reply processor/cumulativetodeltaprocessor/internal/tracking/tracker.go]: done

File: processor/cumulativetodeltaprocessor/internal/tracking/tracker.go
Comment (seankhliao) [reply processor/cumulativetodeltaprocessor/internal/tracking/tracker.go]: ah right, done

File: processor/cumulativetodeltaprocessor/internal/tracking/tracker.go
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/internal/tracking/tracker.go:143]: Looking at this again, it looks like we can get rid of the `keep` bool and do the `MetricType` switch after the `initialValue` switch always since `keep` is never false.

Comment (seankhliao) [reply processor/cumulativetodeltaprocessor/internal/tracking/tracker.go]: nice observation, done

File: processor/cumulativetodeltaprocessor/internal/tracking/tracker.go
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/internal/tracking/tracker.go:146]: nit: we can omit this case.

**Review by TylerHelmuth — APPROVED** [review]

File: processor/cumulativetodeltaprocessor/internal/tracking/tracker.go
Comment (seankhliao) [reply processor/cumulativetodeltaprocessor/internal/tracking/tracker.go]: removed, i originally left it in for completeness

**Comment by seankhliao** [comment]
ack, fixed the panic

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**
